neuraxle/steps/util.py

```diff
@@ -216,17 +235,112 @@ def fit(self, data_inputs: List, expected_outputs: List = None) -> 'StepClonerFo
-        assert len(data_inputs) <= len(self.steps), "Can't have more data_inputs than cloned steps to process them."
+        assert len(data_inputs) >= len(self.steps)
 
         return [self.steps[i].transform(di) for i, di in enumerate(data_inputs)]
 
     def inverse_transform(self, data_output):
-        # As many data outputs than we have cloned steps: each transforms the one it has.
-        assert len(data_output) <= len(self.steps), "Can't have more data_outputs than cloned steps to process them."
+        assert len(data_output) >= len(self.steps)
 
         return [self.steps[i].inverse_transform(di) for i, di in enumerate(data_output)]
 
+    def set_hyperparams(self, hyperparams: HyperparameterSamples) -> BaseStep:
+        self.step = self.step.set_hyperparams(hyperparams.to_flat())
+        return self
+
+    def get_hyperparams(self) -> HyperparameterSamples:
+        return self.step.get_hyperparams()
+
+    def set_hyperparams_space(self, hyperparams_space: HyperparameterSpace) -> 'BaseStep':
+        self.step = self.step.set_hyperparams_space(hyperparams_space.to_flat())
+        return self
+
+    def get_hyperparams_space(self) -> HyperparameterSpace:
+        return self.step.get_hyperparams_space()
+
+    def reverse(self) -> 'BaseStep':
+        """
+        The object will mutate itself such that the ``.transform`` method (and of all its underlying objects
+        if applicable) be replaced by the ``.inverse_transform`` method.
+
+        Note: the reverse may fail if there is a pending mutate that was set earlier with ``.will_mutate_to``.
+
+        :return: a copy of self, reversed. Each contained object will also have been reversed if self is a pipeline.
+        """
+        for step in self.steps:
+            step.mutate(new_method="inverse_transform", method_to_assign_to="transform")
+        return self
+
+    def __reversed__(self) -> 'BaseStep':
+        """
+        The object will mutate itself such that the ``.transform`` method (and of all its underlying objects
+        if applicable) be replaced by the ``.inverse_transform`` method.
+
+        Note: the reverse may fail if there is a pending mutate that was set earlier with ``.will_mutate_to``.
+
+        :return: a copy of self, reversed. Each contained object will also have been reversed if self is a pipeline.
+        """
+        return self.reverse()
+
 
 class DataShuffler:
     pass  # TODO.
+
+
+class OutputTransformerMixin:
+    """
+    Mixin to be able to modify expected_outputs inside a step
+    """
+
+    @abstractmethod
+    def transform_input_output(self, data_inputs, expected_outputs=None) -> Tuple[Any, Any]:
+        """
+        Transform data inputs, and expected outputs at the same time
+
+        :param data_inputs:
+        :param expected_outputs:
+        :return: tuple(data_inputs, expected_outputs)
+        """
+        raise NotImplementedError()
+
+
+class OutputTransformerWrapper(MetaStepMixin, BaseStep):
```

Comment:
we don't need a wrapper anymore. I don't know why we did this at first. it should be a base step with abstractmethod transform_data_inputs_and_expected_outputs only